doc/SAI-Counter-enhancement-hld.md

```diff
@@ -0,0 +1,470 @@
+       gSwitchId,
+       SAI_OBJECT_TYPE_ROUTER_INTERFACE,
+       stats_capability.data());
+
+stats_capability.count = SAI_ROUTER_INTERFACE_STAT_OUT_ERROR_PACKETS+1;
+stats_capability.list = stats.data();
+
+for (uint32_t cnt = 0; cnt < stats_capability.count; cnt++)       
+{
+    supported_counter_ids.push_back(stats_capability.list[cnt]);
+}
+
+status = sai_query_attribute_capability(gSwitchId, SAI_OBJECT_TYPE_ROUTER_INTERFACE, 
+                                        SAI_ROUTER_INTERFACE_ATTR_STAT_ENABLED, 
+                                        &rif_capability);
+if (status == SAI_STATUS_SUCCESS && 
+   (rif_capability.create_implemented || rif_capability.set_implemented))
+{    
+    sai_attribute_t attr;
+    attr.id = SAI_ROUTER_INTERFACE_ATTR_STAT_ENABLED;
+    status = sai_router_intfs_api->get_router_interface_attr(port.m_rif_id,1, &attr); 
+
+    if(status == SAI_STATUS_SUCCESS && attr.value.s32 == SAI_STATS_COUNT_MODE_NONE)
+    {
+        enable = false;
+    }
+    else
+    {
+        /* Return code SAI_STATUS_NOT_IMPLEMENTED if attr not supported. */
+    }
+}
+
+if(enable)
+{
+    status = sai_router_intfs_api->get_router_interface_stats(
+                    rid,
+                    (uint32_t(supported_counter_ids.size()),
+                    (sai_stat_id_t *)supported_counter_ids.data(),
+                    stats.data());
+}
+```
+
+#### Pros 
+- Minimal change to existing work flow.
+#### Cons
+- No Flexiblity to count selective statistics.
+
+### Option 2
+Introduce counter ID attribute that is holding the counter oid of specific statistic enum. 
+Lets take an example of sairouterinterface. Below statistics enums are defined in sairouterinterface,
+```c
+typedef enum _sai_router_interface_stat_t
+{
+    /** Ingress byte stat count */
+    SAI_ROUTER_INTERFACE_STAT_IN_OCTETS,
+
+    /** Ingress packet stat count */
+    SAI_ROUTER_INTERFACE_STAT_IN_PACKETS,
+
+    /** Egress byte stat count */
+    SAI_ROUTER_INTERFACE_STAT_OUT_OCTETS,
+
+    /** Egress packet stat count */
+    SAI_ROUTER_INTERFACE_STAT_OUT_PACKETS,
+
+    /** Byte stat count for packets having errors on router ingress */
+    SAI_ROUTER_INTERFACE_STAT_IN_ERROR_OCTETS,
+
+    /** Packet stat count for packets having errors on router ingress */
+    SAI_ROUTER_INTERFACE_STAT_IN_ERROR_PACKETS,
+
+    /** Byte stat count for packets having errors on router egress */
+    SAI_ROUTER_INTERFACE_STAT_OUT_ERROR_OCTETS,
+
+    /** Packet stat count for packets having errors on router egress */
+    SAI_ROUTER_INTERFACE_STAT_OUT_ERROR_PACKETS
+
+} sai_router_interface_stat_t;
+
+```
+Above enums will be mapped to new counter ID attributes.
+```c
+ /**
+     * @brief Attach a counter
+     *
+     * Ingress byte, packet stat count
+     *
+     * @type sai_object_id_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_COUNTER
+     * @allownull true
+     * @default SAI_NULL_OBJECT_ID
+     */
+    
+    SAI_ROUTER_INTERFACE_ATTR_IN_COUNTER,
```

Comment:
Will this mode be any better if we define something

     * @type sai_object_list_t
     * @flags CREATE_AND_SET
     * @objects SAI_OBJECT_TYPE_COUNTER
     * @allownull true
     * @default SAI_NULL_OBJECT_ID
     */
SAI_ROUTER_INTERFACE_ATTR_COUNTER_IDS

and enforce the list size as SAI_ROUTER_INTERFACE_STAT_MAX and indexing based on the stat ID ?

Reply:
Below are the CONs of using list indexed with stat ID,

- Need to add new stats enum (SAI_XXX_STAT_END)to identify the max number ( saivlan, sairouterinterface, saitunnel).
- Enforce stats id enum to be contiguous.
- How to handle existing disjoint sets in the stat-id enum. Adjusting existing enums to determine MAX_STAT index may not be backward compatible.
For example, in saiport.h, "SAI_PORT_STAT_END" is last enum and debug counters are part of stats enum list. How to arrive the max id in such scenario ? Can we move "SAI_PORT_STAT_END" before "SAI_PORT_STAT_IN_DROP_REASON_RANGE_BASE = 0x00001000". This will break backward compatibility.